omegaconf/_utils.py

```diff
@@ -321,61 +326,35 @@ class ValueKind(Enum):
     VALUE = 0
     MANDATORY_MISSING = 1
     INTERPOLATION = 2
-    STR_INTERPOLATION = 3
 
 
-def get_value_kind(value: Any, return_match_list: bool = False) -> Any:
+def get_value_kind(value: Any) -> ValueKind:
     """
     Determine the kind of a value
     Examples:
-    MANDATORY_MISSING : "???
-    VALUE : "10", "20", True,
-    INTERPOLATION: "${foo}", "${foo.bar}"
-    STR_INTERPOLATION: "ftp://${host}/path"
-
-    :param value: input string to classify
-    :param return_match_list: True to return the match list as well
-    :return: ValueKind
-    """
-
-    key_prefix = r"\${(\w+:)?"
-    legal_characters = r"([\w\.%_ \\/:,-]*?)}"
-    match_list: Optional[List[Match[str]]] = None
-
-    def ret(
-        value_kind: ValueKind,
-    ) -> Union[ValueKind, Tuple[ValueKind, Optional[List[Match[str]]]]]:
-        if return_match_list:
-            return value_kind, match_list
-        else:
-            return value_kind
+    VALUE : "10", "20", True
+    MANDATORY_MISSING : "???"
+    INTERPOLATION: "${foo.bar}", "${foo.${bar}}", "${foo:bar}", "[${foo}, ${bar}]",
+                   "ftp://${host}/path", "${foo:${bar}, [true], {'baz': ${baz}}}"
 
-    from .base import Container
-
-    if isinstance(value, Container):
-        if value._is_interpolation() or value._is_missing():
-            value = value._value()
+    :param value: Input to classify.
+    """
 
     value = _get_value(value)
-    if value == "???":
-        return ret(ValueKind.MANDATORY_MISSING)
-
-    if not isinstance(value, str):
-        return ret(ValueKind.VALUE)
 
-    match_list = list(re.finditer(key_prefix + legal_characters, value))
-    if len(match_list) == 0:
-        return ret(ValueKind.VALUE)
-
-    if len(match_list) == 1 and value == match_list[0].group(0):
-        return ret(ValueKind.INTERPOLATION)
+    if value == "???":
+        return ValueKind.MANDATORY_MISSING
+
+    # We detect potential interpolations simply by the presence of "${" in the string.
+    # This is much more efficient than calling `grammar_parser.parse()`, but one must
+    # be aware that:
+    #   - invalid interpolations are not detected here
+    #   - escaped interpolations (ex: "esc: \${bar}") are identified as interpolations
+    #     (this is actually required to properly un-escape them during resolution)
+    if isinstance(value, str) and "${" in value:
+        return ValueKind.INTERPOLATION
```

Comment:
Can you do a quick benchmark to see the cost of detecting interpolation using this heuristic versus using the parser?
ANTLR is very efficient.

Reply:
That was my first implementation, suddenly the test suite was running much slower.
The comment actually over-simplifies things a bit: calling `grammar_parser.parse()` is not enough, afterwards you also need to visit the parse tree to see if there is any interpolation inside, which brings additional overhead.

Just re-ran a quick test with just `grammar_parser.parse()` (not visiting the parse tree) and the test suite is ~2x slower.

Reply:
> One way to speed it up is to potentially cache the result of is_interpolation per node.
> Now that we have (in master) the beginning of a benchmark suite, you can benchmark it properly.

Just calling `grammar_parser.parse()` multiplies the duration of the benchmark `test_omegaconf_create[large_dict]` by ~4 when using `leaf_value="abc"`. I doubt a cache would help much since in this example there is a large number of nodes.

Is there something in particular you are worried about with the current heuristic? None of the two points mentioned in the comment is a problem, they are here as "FYI" as it may not be obvious. But even if we parsed the expression, it could make sense to keep the same behavior (maybe on the first point we'd want to raise an exception here instead of waiting for someone to access the value, but that's unlikely to matter much).

Reply:
> Easy enough to add a small stupid cache as a node member there there. can you try and see?

Will try later -- no time now

> The detection may have false positives which are later properly evaluated with the parser but no false negatives?

Correct: as soon as there is a `${` in a string, it has to be parsed as an interpolation to be properly evaluated (possibly resulting in an error if the interpolation is not well formed).

Reply:
> > Easy enough to add a small stupid cache as a node member there there. can you try and see?
> 
> Will try later -- no time now

I tried, still a lot slower in the large_dict benchmark -- I think it's because there's a large number of nodes and the overhead comes from checking each node for interpolations, not from re-accessing the interpolation status of nodes multiple times.

Reply:
> Correct: as soon as there is a ${ in a string, it has to be parsed as an interpolation to be properly evaluated (possibly resulting in an error if the interpolation is not well formed).

Then this is indeed fine. resolving.

Reply:
> Than this is indeed fine. resolving.

Cool, FYI I just pushed in beb73da a comment to make it clearer, in case anyone else wonders in the future.